include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,312 @@
+        XTENSOR_ASSERT( weights  .dimension() == 1 );
+        XTENSOR_ASSERT( bin_edges.dimension() == 1 );
+        XTENSOR_ASSERT( weights  .size() == data.size() );
+        XTENSOR_ASSERT( bin_edges.size() >= 2 );
+        // - bin-edges must be sorted
+        XTENSOR_ASSERT( std::is_sorted(bin_edges.cbegin(), bin_edges.cend()) );
+        // - data must be enclosed in the bins
+        XTENSOR_ASSERT( xt::amin(data)[0] >= bin_edges[0] );
+        XTENSOR_ASSERT( xt::amax(data)[0] <= bin_edges[bin_edges.size() - 1] );
+
+        // initialize output
+        xt::xtensor<value_type,1> count = xt::zeros<value_type>({bin_edges.size()-1});
+
+        // indices that sort "data"
+        xt::xtensor<size_t,1> isort = xt::argsort(data);
+
+        // index of the current bin
+        size_type ibin = 0;
+
+        // fill the histogram: loop over (sorted) data
+        for (auto &idx : isort)
+        {
+            // - proceed to the relevant bin
+            while (data[idx] >= bin_edges[ibin+1] && ibin < bin_edges.size()-2)
+            {
+                ++ibin;
+            }
+            // - update the count
+            count[ibin] += weights[idx];
+        }
+
+        // cast type
+        xt::xtensor<double,1> P = xt::cast<double>(count);
+
+        // normalize
+        if ( density )
+        {
+            // - size as doubles
+            double N = static_cast<double>(data.size());
+            // - apply normalization
+            for (size_type i = 0 ; i < bin_edges.size()-1 ; ++i)
+            {
+                P[i] /= ( static_cast<double>(bin_edges[i+1] - bin_edges[i]) * N );
+            }
+        }
+
+        return P;
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bin_edges The bin-edges.
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2>
+    inline auto histogram(
+        E1&& data, E2&& bin_edges, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        return histogram(data, bin_edges, xt::ones<value_type>({data.size()}), density);
+    }
+
+    /**
+     * @ingroup histogram
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param data The data.
+     * @param bins The number of bins. [default: 10]
+     * @param density If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1>
+    inline auto histogram(
+        E1&& data, size_t bins=10, bool density=false
+    )
+    {
+        using value_type = typename std::decay_t<E1>::value_type;
+
+        auto bin_edges = histogram_bin_edges(data, xt::ones<value_type>({data.size()}), bins);
```

Comment:
No forwarding here since `data` is used after, and a potential move here could set it in an unvalid state.

Reply:
@JohanMabille To me this remark seems to apply to the next two occurrences too. Can you comment?